### PR TITLE
Add have a cookie extension

### PR DESCRIPTION
**Pull request by HenriChabrand**
## Description

Add "Have a Cookie" extension, that allows Google Chrome users to easily retrieve cookies from any visited websites.

## Type of change

- New extension 

## Screencast

Screenshot 2021-10-31 at 23 15 06
Screenshot 2021-10-31 at 23 15 30

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I checked that extension's icon has size of `512x512`

### CODE REVIEW

**Comment by HenriChabrand** [comment]
@PitNikola think for LinkedIn – I've reduced the need to have permanent access to the Keychain.

Here the user need to "Allow" once.

File: extensions/have-a-cookie/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,240 @@
+import { 
+  ActionPanel, 
+  PushAction, 
+  CopyToClipboardAction, 
+  Detail,
```
Comment (pernielsentikaer) [inline extensions/have-a-cookie/src/index.tsx:5]:
```suggestion
```

File: extensions/have-a-cookie/src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,240 @@
+import { 
+  ActionPanel, 
+  PushAction, 
+  CopyToClipboardAction, 
+  Detail, 
+  PasteAction, 
+  List, 
+  OpenInBrowserAction, 
+  showToast, 
+  ToastStyle ,
```
Comment (pernielsentikaer) [inline extensions/have-a-cookie/src/index.tsx:10]:
```suggestion
```

File: extensions/have-a-cookie/src/util/chrome-cookie-helper.js [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+ int = require('int'),
+ url = require('url'),
+ crypto = require('crypto'),
+ os = require('os'),
+ fs = require('fs'),
+ dpapi,
+ Cookie = tough.Cookie,
+ path,
+ ITERATIONS,
+ dbClosed = false;
```
Comment (pernielsentikaer) [inline extensions/have-a-cookie/src/util/chrome-cookie-helper.js:13]: Is this in use?

File: extensions/have-a-cookie/src/util/chrome-cookie-helper.js [hunk 4]
```diff
@@ -0,0 +1,523 @@
+ var tld = require('tldjs'),
+ tough = require('tough-cookie'),
+ request = require('request'),
+ int = require('int'),
+ url = require('url'),
+ crypto = require('crypto'),
+ os = require('os'),
+ fs = require('fs'),
+ dpapi,
+ Cookie = tough.Cookie,
```
Comment (pernielsentikaer) [inline extensions/have-a-cookie/src/util/chrome-cookie-helper.js:10]: Is this in use?

File: extensions/have-a-cookie/src/util/chrome-cookie-helper.js [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+
+	decoded = decoded.toString('utf8');
+
+	return decoded;
+
+}
+
+async function getDerivedKey(callback) {
+
+	var keytar,
```
Comment (pernielsentikaer) [inline extensions/have-a-cookie/src/util/chrome-cookie-helper.js:58]: Is this in use?

File: extensions/have-a-cookie/src/util/chrome-cookie-helper.js [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+		return callback('Only Mac, Windows, and Linux are supported.');
+
+	}
+
+	
+
+	if (format instanceof Function) {
+		callback = format;
+		format = null;
+	}
```
Comment (pernielsentikaer) [inline extensions/have-a-cookie/src/util/chrome-cookie-helper.js:265]: This code is also at line 420-437 - could maybe be a new function to keep it dry

File: extensions/have-a-cookie/src/util/chrome-cookie-helper.js [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+
+						} else if (encryptedValue[0] == 0x76 && encryptedValue[1] == 0x31 && encryptedValue[2] == 0x30 ){
+							localState = JSON.parse(fs.readFileSync(os.homedir() + '/AppData/Local/Google/Chrome/User Data/Local State'));
+							b64encodedKey = localState.os_crypt.encrypted_key;
+							encryptedKey = new Buffer.from(b64encodedKey,'base64');
+							key = dpapi.unprotectData(encryptedKey.slice(5, encryptedKey.length), null, 'CurrentUser');
+							nonce = encryptedValue.slice(3, 15);
+							tag = encryptedValue.slice(encryptedValue.length - 16, encryptedValue.length);
+							encryptedValue = encryptedValue.slice(15, encryptedValue.length - 16);
+							cookie.value = decryptAES256GCM(key, encryptedValue, nonce, tag).toString('utf-8');
```
Comment (pernielsentikaer) [inline extensions/have-a-cookie/src/util/chrome-cookie-helper.js:331]: Should be let or const

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @HenriChabrand 👋 

I just gave your extension a quick spin and it looks very nice 💯 💪 

There is a lot going on in the chrome-cookie-helper and I just commented on a few things that could be corrected, overall it seems to do the job as it should even though I'm not completely sure what's going on everywhere 😄

File: extensions/have-a-cookie/src/util/chrome-cookie-helper.js (hunk 6 above)
Comment (HenriChabrand) [reply extensions/have-a-cookie/src/util/chrome-cookie-helper.js:265]: Actually as Raycast is only available on Mac, there is no need to have this here

Comment (pernielsentikaer) [reply extensions/have-a-cookie/src/util/chrome-cookie-helper.js]: Correct :)

**Comment by HenriChabrand** [comment]
Ahah yes, it's a bit complex 😅 

The source code inspiring `chrome-cookie-helper` comes from this repo: [chrome-cookies-secure](https://github.com/bertrandom/chrome-cookies-secure)
I had to tweak it a lot to make it works as Raycast doesn’t officially support compiled node modules.

I cleaned it as much as I could – let me know if that's fine for you 👍

And thanks a lot for your feedback @pernielsentikaer 🙌

File: extensions/have-a-cookie/package.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "have-a-cookie",
+  "title": "Have a Cookie",
+  "description": "Retrieves any cookie from Google Chrome.",
```
Comment (pernielsentikaer) [inline extensions/have-a-cookie/package.json:4]:
```suggestion
  "description": "Search and retrieve any cookies from Google Chrome",
```

I think I would do this, but again, not critically.

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @HenriChabrand 👋 

I think we're ready to merge this, I left a quickie that could be corrected

I will check in a few hours when I'm available again 💯

**Comment by PitNikola** [comment]
@HenriChabrand, as we discussed in Slack, there are security concerns with letting this extension requesting keychain access to Chrome Web Storage. Closing this PR for now, but feel free to publish it in your own repo, maybe some folks would find it useful. Thanks for your understanding!

**Pull request closed without merging**